release-23.1: roachtest: backup-restore/mixed-version updates

renatolabs opened the pull request:
Backport:
  * Parts of ee93d72 (see thread).
  * 1/1 commits from "roachtest: backup-restore/mixed-version use system db when querying job status" (#110208)
  * 1/1 commits from "roachtest: add round trip backup restore test" (#108372)
  * 1/1 commits from "mixedversion: replace LowestBinaryVersion with ClusterVersionAtLeast" (#111912)
  * 1/1 commits from "roachtest: check cluster version after resetting cluster" (#112193)

Please see individual PRs for details.

/cc @cockroachdb/release

Release justification: test-only changes (+ small bug fix in function not previously used).

Commit:
roachtest: add round trip backup restore test

Add a round trip backup and restore roachtest. This test runs several backups
and backup verifications while a cluster is under load.

The patch also refactors the backup and backup verification logic introduced
in the backup-restore/mixed-version test into a separate
`BackupRestoreTestDriver` struct so that they can be used by both the mixed
version test and the new round trip test.

Release note: None

Commit:
roachtest: backup-restore/mixed-version use system db when querying job status

Previously in `waitForJobSuccess()` in the backup-restore/mixed-version test,
if the job was a full cluster restore that dropped `defaultdb`, the query to
the jobs table would fail with `pq: database "defaultdb" is offline: restoring`
as our connection was to `defaultdb`, which is dropped during full cluster
restore. We fix this by adding an option to connect to a different database
in roachtests and connecting to the system database in `waitForJobSuccess()`
instead.

Fixes #110165

Release note: None

Commit:
mixedversion: replace LowestBinaryVersion with ClusterVersionAtLeast

This removes the `LowestBinaryVersion` helper function and replaces it
with a new `ClusterVersionAtLeast` function. This change shifts the
focus from released binaries versions to cluster versions. At the end
of the day, upgrade tests are much more concerned about cluster
versions than binary versions, so orienting tests around that concept
is both more intuitive and less error prone.

Epic: none

Release note: None

Commit:
roachtest: check cluster version after resetting cluster

This fixes the `backup-restore/mixed-version` test to perform cluster
version checks _after_ the cluster has been potentially reset while
verifying cluster backups. Previously, we were checking the version
before, which could lead to wrongly using features that were not
available yet.

Fixes: #112189.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/112256)
<!-- Reviewable:end -->

srosenberg commented:
The unit tests are failing because the fix [1] for `AtLeast` wasn't backported into 23.1.

[1] https://github.com/cockroachdb/cockroach/commit/ee93d72ab145eb7e8697b4dc9edd1f30dca1856e

renatolabs commented:
Good catch! Interesting that `AtLeast` was broken in that way. 

I could backport just the fix to `AtLeast` in this PR, but I'm also curious about #102263 in general; it changes something based on an internal version between 22.2 and 23.1, but it wasn't backported. @RaduBerinde why was it not backported?

RaduBerinde commented:
The features in that PR were not enabled at all in 23.1 so it wasn't important to backport. There seemed to be no other uses of AtLeast either.

renatolabs commented:
> The features in that PR were not enabled at all in 23.1

Could you elaborate, please? What features? The commit above has logic that changes behaviour based on whether `22.2-64` is active, so I'd expect it to matter on release-23.1.

In any case, I'll just backport the fix for `AtLeast()` in this PR.

RaduBerinde commented:
The Creator ID is used for disaggeegated storage which is not enabled in any form in 23.1. It does not matter if we initialize the ID or not in 23.1. The 22.2.64 version is necessary we to set it, but if we set it or not is inconsequential for 23.1.

Commit:
version: fix implementation of `AtLeast` function

This is a partial cherry-pick of ee93d72, without the pebble change to
set the creator ID, unnecessary in 23.1.

Epic: none

Release note: None

renatolabs commented:
Pushed a partial application of ee93d72, including the fix for `AtLeast` and tests, but without the Pebble changes.

RaduBerinde commented:
AtLeast fix LGTM
